include/ginkgo/core/matrix/fbcsr.hpp

```diff
@@ -0,0 +1,390 @@
+
+    void convert_to(Dense<ValueType> *other) const override;
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void convert_to(Csr<ValueType, IndexType> *result) const override;
+
+    void move_to(Csr<ValueType, IndexType> *result) override;
+
+    /**
+     * Get the block sparsity pattern in CSR-like format
+     *
+     * Note that the actual non-zero values are never copied;
+     * the result always has a value array of size 1 with the value 1.
+     */
+    void convert_to(SparsityCsr<ValueType, IndexType> *result) const override;
+
+    void move_to(SparsityCsr<ValueType, IndexType> *result) override;
+
+    /** Convert COO data into block CSR
+     *
+     * @warning Unlike Csr::read, here explicit non-zeros are NOT dropped.
+     */
+    void read(const mat_data &data) override;
+
+    void write(mat_data &data) const override;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    std::unique_ptr<Diagonal<ValueType>> extract_diagonal() const override;
+
+    std::unique_ptr<absolute_type> compute_absolute() const override;
+
+    void compute_absolute_inplace() override;
+
+    /**
+     * Sorts all (value, col_idx) pairs in each row by column index
+     */
+    void sort_by_column_index();
+
+    /*
+     * Tests if all row entry pairs (value, col_idx) are sorted by column index
+     *
+     * @returns True if all row entry pairs (value, col_idx) are sorted by
+     *          column index
+     */
+    bool is_sorted_by_column_index() const;
+
+    /**
+     * @return The values of the matrix.
+     */
+    value_type *get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @see Fbcsr::get_const_values()
+     */
+    const value_type *get_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * @copydoc Fbcsr::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * @return The column indexes of the matrix.
+     */
+    index_type *get_col_idxs() noexcept { return col_idxs_.get_data(); }
+
+    /**
+     * @see Fbcsr::get_const_col_idxs()
+     */
+    const index_type *get_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
+    }
+
+    /**
+     * @copydoc Fbcsr::get_col_idxs()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
```

Comment:
If the user uses `get_values` on an `Fbcsr` object, the outcome depends on whether the object is const, and whether they assign the returned pointer to a const. For const `Fbcsr` objects, the returned pointer is const, otherwise it depends on the type to which the return value is assigned. For an `auto` return variable, it uses the non-const version if possible. I think this is nice to have and will always work as the user intends. But if people have objections I can remove it.